Fix invalid "No more buffers already set" in Broadcast buffer

dain opened the pull request:
When the broadcast buffer is closed early the buffer may not have received
the final buffer declaration, so buffer creation must be allowed in this case.

This causes queries to fail with the following stack:
```
java.lang.IllegalStateException: No more buffers already set
    at com.google.common.base.Preconditions.checkState(Preconditions.java:173)
    at com.facebook.presto.execution.buffer.BroadcastOutputBuffer.getBuffer(BroadcastOutputBuffer.java:301)
    at com.facebook.presto.execution.buffer.BroadcastOutputBuffer.get(BroadcastOutputBuffer.java:246)
    at com.facebook.presto.execution.buffer.LazyOutputBuffer.get(LazyOutputBuffer.java:177)
    at com.facebook.presto.execution.SqlTask.getTaskResults(SqlTask.java:341)
    at com.facebook.presto.execution.SqlTaskManager.getTaskResults(SqlTaskManager.java:335)
    at com.facebook.presto.server.TaskResource.getResults(TaskResource.java:242)
```

haozhun commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/BroadcastOutputBuffer.java:
```diff
@@ -162,19 +164,26 @@ public void setOutputBuffers(OutputBuffers newOutputBuffers)
                 this.state.compareAndSet(OPEN, NO_MORE_BUFFERS);
                 this.state.compareAndSet(NO_MORE_PAGES, FLUSHING);
+
+                pagesToDereference = ImmutableList.copyOf(initialPagesForNewBuffers);
```
Note to self: this effectively moves `noMoreBuffers` to the same synchronized block.

haozhun commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/BroadcastOutputBuffer.java:
```diff
@@ -263,9 +275,17 @@ public void setNoMorePages()
+                pages = ImmutableList.copyOf(initialPagesForNewBuffers);
+                initialPagesForNewBuffers.clear();
+            }
+            pages.forEach(SerializedPageReference::dereferencePage);
```
Note to self: this effectively inlined `noMoreBuffers` but removed check that all created buffers are declared.

haozhun commented on presto-main/src/main/java/com/facebook/presto/execution/buffer/BroadcastOutputBuffer.java:
```diff
@@ -290,7 +310,10 @@ private synchronized ClientBuffer getBuffer(OutputBufferId id)
+        // NOTE: buffers are allowed to be created in the FINISHED state because destroy() can move to the finished state
+        // without a clean "no-more-buffers" message from the scheduler.  This happens with limit queries and is ok because
+        // the buffer will be immediately destroyed.
+        checkState(state.get().canAddBuffers() || state.get() == FINISHED, "No more buffers already set");
```
Note to self: this allows creating arbitrary buffers when in finished state.

dain commented:
@haozhun I have a new approach to this problem.  I reverted the parts that removed `noMoreBuffers` and updated it to only verify all created buffers have been declared if `outputBuffers.isNoMoreBufferIds()` is true.

I have also update `ArbitraryOutputBuffer`.
